File: ios/engine/StreamHub.swift
// StreamHub.swift
// BridgeKit iOS engine — multiplexes one provider AsyncStream across N consumers.
//
// Port of io/github/malopezr7/bridgekit/core/StreamHub.kt
//
// DESIGN (Decision 2): AsyncStream is single-consumer; Kotlin MutableSharedFlow is
// multi-consumer. Port = hand-written multicast: HubEntry holds a dictionary of
// consumer continuations (UUID → onNext closure) guarded by the engine lock.
// One upstream Task runs `for try await v in upstream` and fans out to all consumers.
//
// H-5: terminalResult stored BEFORE fan-out. A late consumer that finds terminalResult
// non-nil delivers the terminal INLINE (not via replay) — Swift delivers synchronously
// without needing an UNDISPATCHED watcher.
//
// H-6: identity-checked remove: `if hubs[key] === entry { removeValue }`.
//
// H-7 lives in Router.openStream (recheck isLive after streamJobs insert).
//
// INV-3 (H-5), INV-4 (H-6) implemented here.
//
// PORT NOTE: Kotlin uses MutableSharedFlow for fan-out + CompletableDeferred for
// terminal. Swift equivalent: per-consumer closures dictionary + terminalResult?
// stored under the engine lock. No additional concurrency primitives needed.

import Foundation

// ---- Hub sentinels ---------------------------------------------------------

/// Sentinel: upstream completed normally.
internal final class HubTerminalOK {}
internal let HUB_TERMINAL_OK = HubTerminalOK()

/// Upstream completed with an error.
internal struct HubTerminalError {
    let message: String
}

/// Union type for a hub terminal signal.
internal enum HubTerminal {
    case ok
    case error(String)
}

// ---- StreamHub -------------------------------------------------------------

internal final class StreamHub {

    private struct HubKey: Hashable {
        let contractId: String
        let member: String
        let scopeKey: String
        let paramsHash: Int64
    }

    /// State for one active upstream.
    private final class HubEntry {
        /// Per-consumer callbacks keyed by UUID.
        /// onNext receives the raw value (Any?); terminal delivery is separate.
        var consumers: [UUID: (Any?) -> Void] = [:]
        var onEndCallbacks: [UUID: ([String: Any?]) -> Void] = [:]
        var terminalFiredFor: Set<UUID> = []

        var upstreamTask: Task<Void, Never>? = nil

        /// INV-3 (H-5): set BEFORE fan-out. Late consumer checks this under the lock.
        var terminalResult: HubTerminal? = nil

        var replayLatest: Any?? = nil  // nil = no replay; .some(nil) = last was nil
        var hasReplay: Bool = false
    }

    // All hub state is guarded by the engine NSRecursiveLock passed at init.
    // PORT NOTE: Kotlin used ConcurrentHashMap — in Swift the engine lock replaces it.
    private var hubs: [HubKey: HubEntry] = [:]

    /// Weak reference to the engine lock. StreamHub borrows the engine lock for all
    /// mutations so the lock hierarchy is: engineLock → streamHub state.
    /// This is safe because StreamHub is only ever accessed from within the engine lock
    /// on sync paths, and the upstream Task fans out OUTSIDE the lock.
    private unowned(unsafe) let lock: NSRecursiveLock

    // PORT NOTE: Kotlin used `CoroutineScope(engineScope)`. Swift uses unstructured
    // Tasks that inherit no scope. Tasks are stored in HubEntry.upstreamTask and
    // cancelled explicitly on cancelAll() / last-consumer-detach.
    internal init(lock: NSRecursiveLock) {
        self.lock = lock
    }

    // -------------------------------------------------------------------------
    // attach
    // -------------------------------------------------------------------------

    /// Attach a consumer to the shared stream for (contractId, member, scope, paramsHash).
    ///
    /// Returns a `Task<Void, Never>` that the caller stores in streamPumpJobs. Cancelling
    /// it detaches this consumer from the hub and releases the upstream when the last
    /// consumer leaves.
    ///
    /// Called UNDER the engine lock (Router holds the lock during openStream).
    @discardableResult
    internal func attach(
        contractId: String,
        member: String,
        scope: Scope,
        paramsHash: Int64,
        openStream: @escaping () -> AsyncThrowingStream<Any?, Error>,
        latestOnly: Bool = false,
        sticky: Bool = false,
        onNext: @escaping ([String: Any?]) -> Void,
        onEnd: @escaping ([String: Any?]) -> Void
    ) -> Task<Void, Never> {
        let key = HubKey(
            contractId: contractId,
            member: member,
            scopeKey: scope.serialized(),
            paramsHash: paramsHash
        )
        let consumerID = UUID()

        // getOrCreate the HubEntry — first caller starts the upstream.
        let entry: HubEntry
        if let existing = hubs[key] {
            entry = existing
        } else {
            let newEntry = HubEntry()
            hubs[key] = newEntry
            entry = newEntry
        }

        // Register consumer before starting upstream to avoid missing early emissions.
        entry.consumers[consumerID] = { [weak entry] value in
            // Called OUTSIDE the engine lock by the upstream Task.
            // Check if terminal already fired for this consumer (atomic gate via entry state).
            guard let entry = entry else { return }
            onNext(["v": value])
            // Store replay if needed
            if latestOnly || sticky {
                entry.replayLatest = .some(value)
                entry.hasReplay = true
            }
        }
        entry.onEndCallbacks[consumerID] = onEnd

        // INV-3 (H-5): if upstream already terminated, deliver terminal inline to this
        // late consumer — no replay needed; terminal stored in terminalResult.
        if let terminal = entry.terminalResult {
            entry.terminalFiredFor.insert(consumerID)
            // Deliver inline (caller is under lock; we schedule a Task to call onEnd
            // off-lock to avoid reentrancy).
            // PORT NOTE: Kotlin UNDISPATCHED watcher — Swift equivalent: async Task that
            // starts immediately. We release the lock by scheduling as async Task.
            let terminalEnv = Self.terminalEnv(terminal, contractId: contractId, member: member, scope: scope)
            entry.consumers.removeValue(forKey: consumerID)
            entry.onEndCallbacks.removeValue(forKey: consumerID)
            // Return a task that immediately fires onEnd.
            return Task {
                onEnd(terminalEnv)
            }
        }

        // replay=1: deliver last known value to late subscriber.
        if (latestOnly || sticky) && entry.hasReplay {
            if let replayVal = entry.replayLatest {
                onNext(["v": replayVal])
            } else {
                onNext(["v": Optional<Any>.none as Any?])
            }
        }

        // Start the upstream Task exactly once (for the first attaching consumer).
        if entry.upstreamTask == nil || entry.upstreamTask?.isCancelled == true {
            let capturedKey = key
            entry.upstreamTask = Task {
                do {
                    let stream = openStream()
                    for try await value in stream {
                        // Fan out OUTSIDE the engine lock.
                        // Snapshot consumers under a brief lock acquisition.
                        let snapshot: [(UUID, (Any?) -> Void)]
                        self.lock.lock()
                        snapshot = Array(self.hubs[capturedKey]?.consumers ?? [:])
                        // Store replay
                        if let e = self.hubs[capturedKey] {
                            if latestOnly || sticky {
                                e.replayLatest = .some(value)
                                e.hasReplay = true
                            }
                        }
                        self.lock.unlock()
                        for (_, cb) in snapshot {
                            cb(value)
                        }
                    }
                    // Normal completion.
                    // INV-3 (H-5): set terminalResult BEFORE fan-out.
                    self.lock.lock()
                    if let e = self.hubs[capturedKey] {
                        e.terminalResult = .ok
                    }
                    let endSnapshot: [(UUID, ([String: Any?]) -> Void)]
                    let firedSnapshot: Set<UUID>
                    if let e = self.hubs[capturedKey] {
                        endSnapshot = Array(e.onEndCallbacks)
                        firedSnapshot = e.terminalFiredFor
                        e.consumers.removeAll()
                        e.onEndCallbacks.removeAll()
                    } else {
                        endSnapshot = []
                        firedSnapshot = []
                    }
                    // INV-4 (H-6): identity-checked remove.
                    if let e = self.hubs[capturedKey], self.hubs[capturedKey] === e {
                        self.hubs.removeValue(forKey: capturedKey)
                    }
                    self.lock.unlock()
                    // Fan out terminal OUTSIDE lock.
                    let termEnv: [String: Any?] = ["ok": true, "value": nil]
                    for (id, cb) in endSnapshot where !firedSnapshot.contains(id) {
                        cb(termEnv)
                    }
                } catch is CancellationError {
                    // Cooperative cancel — no terminal emitted. Just clean up.
                    self.lock.lock()
                    // INV-4 (H-6): identity-checked remove.
                    if let e = self.hubs[capturedKey], self.hubs[capturedKey] === e {
                        self.hubs.removeValue(forKey: capturedKey)
                    }
                    self.lock.unlock()
                } catch {
                    // Upstream error.
                    // INV-3 (H-5): set terminalResult BEFORE fan-out.
                    let msg = error.localizedDescription
                    self.lock.lock()
                    if let e = self.hubs[capturedKey] {
                        e.terminalResult = .error(msg)
                    }
                    let endSnapshot: [(UUID, ([String: Any?]) -> Void)]
                    let firedSnapshot: Set<UUID>
                    if let e = self.hubs[capturedKey] {
                        endSnapshot = Array(e.onEndCallbacks)
                        firedSnapshot = e.terminalFiredFor
                        e.consumers.removeAll()
                        e.onEndCallbacks.removeAll()
                    } else {
                        endSnapshot = []
                        firedSnapshot = []
                    }
                    // INV-4 (H-6): identity-checked remove.
                    if let e = self.hubs[capturedKey], self.hubs[capturedKey] === e {
                        self.hubs.removeValue(forKey: capturedKey)
                    }
                    self.lock.unlock()
                    let termEnv: [String: Any?] = [
                        "ok": false,
                        "code": "PROVIDER_ERROR",
                        "message": msg,
                        "contractId": contractId,
                        "member": member,
                        "scope": ["kind": scope.serialized()]
                    ]
                    for (id, cb) in endSnapshot where !firedSnapshot.contains(id) {
                        cb(termEnv)
                    }
                }
            }
        }

        // Return a Task that represents this consumer's lifetime.
        // Cancelling this Task detaches the consumer.
        let capturedKey = key
        let consumerTask = Task<Void, Never> {
            await withTaskCancellationHandler {
                // Body: suspend until cancelled. The real work happens via the
                // upstream Task's fan-out into the onNext closure registered above.
                // PORT NOTE: Kotlin consumerJob.invokeOnCompletion { detach() }.
                // Swift: use withTaskCancellationHandler to detach on cancel.
                try? await Task.sleep(nanoseconds: .max)
            } onCancel: {
                self.lock.lock()
                if let e = self.hubs[capturedKey] {
                    e.consumers.removeValue(forKey: consumerID)
                    e.onEndCallbacks.removeValue(forKey: consumerID)
                    // If last consumer, cancel upstream.
                    if e.consumers.isEmpty {
                        e.upstreamTask?.cancel()
                        // INV-4 (H-6): identity-checked remove.
                        if self.hubs[capturedKey] === e {
                            self.hubs.removeValue(forKey: capturedKey)
                        }
                    }
                }
                self.lock.unlock()
            }
        }

        return consumerTask
    }

    // -------------------------------------------------------------------------
    // detach (explicit, for Hub-internal use)
    // -------------------------------------------------------------------------

    /// Decrement refcount and cancel upstream when the last consumer leaves.
    /// Called from consumerTask.onCancel handler above.
    internal func detach(contractId: String, member: String, scope: Scope, paramsHash: Int64, consumerID: UUID) {
        let key = HubKey(
            contractId: contractId,
            member: member,
            scopeKey: scope.serialized(),
            paramsHash: paramsHash
        )
        guard let entry = hubs[key] else { return }
        entry.consumers.removeValue(forKey: consumerID)
        entry.onEndCallbacks.removeValue(forKey: consumerID)
        if entry.consumers.isEmpty {
            entry.upstreamTask?.cancel()
            // INV-4 (H-6): identity-checked remove.
            if hubs[key] === entry { hubs.removeValue(forKey: key) }
        }
    }

    // -------------------------------------------------------------------------
    // cancelAll (epoch swap)
    // -------------------------------------------------------------------------

    /// Cancel all active hub entries — called on epoch swap (INV-2 companion).
    internal func cancelAll() {
        for (_, entry) in hubs {
            entry.upstreamTask?.cancel()
        }
        hubs.removeAll()
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private static func terminalEnv(
        _ terminal: HubTerminal,
        contractId: String,
        member: String,
        scope: Scope
    ) -> [String: Any?] {
        switch terminal {
        case .ok:
            return ["ok": true, "value": nil]
        case .error(let msg):
            return [
                "ok": false,
                "code": "PROVIDER_ERROR",
                "message": msg,
                "contractId": contractId,
                "member": member,
                "scope": ["kind": scope.serialized()]
            ]
        }
    }
}

// ---- AsyncStream.map extension (INV-9 support for L5 generated outbound) ---

// PORT NOTE: The L5 generated Swift outbound client calls `.map { }` on AsyncStream
// values. `AsyncStream.map` is NOT in the Swift stdlib. This extension provides it.
// Placed here (engine layer) so the runtime and generated code can use it without
// importing NitroModules.
extension AsyncStream {
    /// Returns a new AsyncStream whose elements are the results of applying the
    /// given transform to each element of this stream.
    public func map<T>(_ transform: @escaping (Element) async -> T) -> AsyncStream<T> {
        var iterator = makeAsyncIterator()
        return AsyncStream<T> { continuation in
            Task {
                while let element = await iterator.next() {
                    let transformed = await transform(element)
                    continuation.yield(transformed)
                }
                continuation.finish()
            }
        }
    }
}

extension AsyncThrowingStream {
    /// Returns a new AsyncThrowingStream whose elements are the results of applying
    /// the given transform to each element of this stream.
    public func map<T>(_ transform: @escaping (Element) async throws -> T) -> AsyncThrowingStream<T, Error> {
        var iterator = makeAsyncIterator()
        return AsyncThrowingStream<T, Error> { continuation in
            Task {
                do {
                    while let element = try await iterator.next() {
                        let transformed = try await transform(element)
                        continuation.yield(transformed)
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }
}

File: packages/core/android/src/main/java/com/bridgekit/core/StreamHub.kt
package com.bridgekit.core

import com.bridgekit.runtime.InboundContractAdapter
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.buffer
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.launch
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

// StreamHub — multiplexes one provider Flow across N consumers keyed by
// (contractId, member, scope, paramsHash).
//
// Shares ONE provider invocation across all consumers with identical params.
// A second consumer subscribing while one is already running attaches to the
// existing SharedFlow fan-out; the upstream Flow is cancelled only when the
// last consumer detaches (refcount).
//
// `latestOnly` or `sticky` descriptors use replay = 1 on the SharedFlow
// so a late subscriber immediately receives the last emitted item.

internal class StreamHub(
    private val engineScope: CoroutineScope,
) {
    /** Capacity of the upstream buffer before DROP_OLDEST. Mirrors Router STREAM_BUFFER_CAPACITY. */
    private val UPSTREAM_BUFFER = 64

    private data class HubKey(
        val contractId: String,
        val member: String,
        val scopeKey: String,
        val paramsHash: Long,
    )

    private inner class HubEntry(
        val sharedFlow: MutableSharedFlow<Any?>,
        val refCount: AtomicInteger = AtomicInteger(0),
        var upstreamJob: Job? = null,
        // terminalDeferred completes exactly once when upstream terminates (ok or error).
        // Used to detect "already completed" state and deliver terminal to late consumers.
        // CompletableDeferred.await() resolves the same value to all awaiting callers
        // regardless of when they call await() relative to complete().
        val terminalDeferred: CompletableDeferred<Any?> = CompletableDeferred(),
    )

    private val hubs = ConcurrentHashMap<HubKey, HubEntry>()

    /**
     * Attach a consumer to the shared stream for (contractId, member, scope, paramsHash).
     *
     * If no hub entry exists for this key, starts the upstream provider Flow and fans it
     * out via a MutableSharedFlow. On detach (returned lambda), the refcount is decremented;
     * when it reaches 0, the upstream Job is cancelled and the entry is removed.
     *
     * @param contractId  The contract identifier.
     * @param member      The stream member name.
     * @param scope       The binding scope.
     * @param paramsHash  Deterministic hash of encoded params (use stableHash key-sorting).
     * @param adapter     The inbound adapter for this binding (provides the upstream Flow).
     * @param payload     The raw params map forwarded to the adapter's openStream.
     * @param streamEpoch The epoch at which this stream was opened.
     * @param streamId    Unique stream ID for epoch tracking in streamPumpJobs.
     * @param latestOnly  If true, replay = 1 (late subscriber gets last value).
     * @param sticky      Alias for latestOnly.
     * @param onNext      JS consumer callback for each item.
     * @param onEnd       JS consumer callback on stream termination.
     * @return the consumer collection Job. Cancelling it (e.g. Router.closeStream,
     *         epoch swap, or scope cancellation) automatically detaches this consumer
     *         from the hub, releasing the upstream when the last consumer leaves.
     */
    fun attach(
        contractId: String,
        member: String,
        scope: Scope,
        paramsHash: Long,
        adapter: InboundContractAdapter,
        payload: Map<String, Any?>?,
        streamEpoch: Long,
        streamId: String,
        latestOnly: Boolean = false,
        sticky: Boolean = false,
        onNext: (Map<String, Any?>) -> Unit,
        onEnd: (Map<String, Any?>) -> Unit,
    ): Job {
        val key = HubKey(contractId, member, scope.serialize(), paramsHash)
        val replay = if (latestOnly || sticky) 1 else 0

        // getOrCreate hub entry — first caller starts the upstream, rest join.
        // hubs.remove uses value-aware remove(key, entry) so a new entry for the same
        // key (registered after this entry completes) is never accidentally evicted.
        val entry = hubs.getOrPut(key) {
            val shared = MutableSharedFlow<Any?>(
                replay = replay,
                extraBufferCapacity = UPSTREAM_BUFFER,
                onBufferOverflow = BufferOverflow.DROP_OLDEST,
            )
            HubEntry(sharedFlow = shared)
        }

        entry.refCount.incrementAndGet()

        // Start the upstream Flow exactly once (for the first attaching consumer).
        // The try/catch classifies completion as exactly one terminal — HubTerminalError
        // on failure, HUB_TERMINAL_OK on normal completion — and rethrows
        // CancellationException for cooperative cancel.
        synchronized(entry) {
            if (entry.upstreamJob == null || entry.upstreamJob?.isActive == false) {
                entry.upstreamJob = engineScope.launch {
                    try {
                        val flow = adapter.openStream(member, payload)
                        flow
                            .buffer(UPSTREAM_BUFFER, BufferOverflow.DROP_OLDEST)
                            // No .catch here — errors must reach the outer catch so we can
                            // emit HubTerminalError instead of silently emitting HUB_TERMINAL_OK.
                            .collect { value -> entry.sharedFlow.emit(value) }
                        // Upstream completed normally.
                        // Complete terminalDeferred before SharedFlow emit so any consumer
                        // awaiting the deferred sees it atomically with the SharedFlow emission.
                        entry.terminalDeferred.complete(HUB_TERMINAL_OK)
                        entry.sharedFlow.emit(HUB_TERMINAL_OK)
                    } catch (ce: CancellationException) {
                        // Cooperative cancel (epoch-swap, last-consumer-detach). Do NOT emit
                        // any terminal sentinel — the consumer job is already being cancelled.
                        throw ce
                    } catch (e: Throwable) {
                        // Upstream error — complete terminalDeferred with error terminal.
                        val errorTerminal = HubTerminalError(e.message ?: "Stream error")
                        entry.terminalDeferred.complete(errorTerminal)
                        entry.sharedFlow.emit(errorTerminal)
                    } finally {
                        // Value-aware remove — only remove THIS entry, not a live replacement
                        // registered under the same key by a new provider after this one completed.
                        hubs.remove(key, entry)
                    }
                }
            }
        }

        // Early-exit for late consumers: if the upstream already terminated
        // (terminalDeferred is complete), the terminal emission was already dropped by
        // SharedFlow (replay=0). We must deliver it directly via a watcher, not SharedFlow.
        // For ongoing streams (deferred NOT complete) we skip the watcher — sharedFlow.collect
        // handles terminal delivery with no extra overhead.

        val terminalFired = AtomicBoolean(false)

        lateinit var consumerJob: Job
        consumerJob = engineScope.launch {
            // Subscribe to sharedFlow immediately — no child launches before collect.
            // With replay=0 any emission before subscription is registered will be dropped;
            // starting collect first minimises that window.
            try {
                entry.sharedFlow.collect { value ->
                    when {
                        value === HUB_TERMINAL_OK -> {
                            if (terminalFired.compareAndSet(false, true)) {
                                onEnd(mapOf("ok" to true, "value" to null))
                            }
                            // Cancel this consumer's coroutine so the SharedFlow collection
                            // actually stops. invokeOnCompletion below will then call detach().
                            consumerJob.cancel(CancellationException("terminal-ok"))
                        }
                        value is HubTerminalError -> {
                            if (terminalFired.compareAndSet(false, true)) {
                                onEnd(mapOf(
                                    "ok" to false,
                                    "code" to "PROVIDER_ERROR",
                                    "message" to value.message,
                                    "contractId" to contractId,
                                    "member" to member,
                                    "scope" to mapOf("kind" to scope.serialize()),
                                ))
                            }
                            // Cancel so the collector stops — guarantees exactly one terminal.
                            consumerJob.cancel(CancellationException("terminal-error"))
                        }
                        else -> {
                            onNext(mapOf("v" to value))
                        }
                    }
                }
            } finally {
                // Watcher cleanup (if any) handled via invokeOnCompletion below.
            }
        }

        // Launch a watcher ONLY if the upstream has already completed.
        // UNDISPATCHED: starts on the current thread, suspends immediately at await()
        // (which returns instantly since isCompleted is true), then resumes on Default.
        var watcherJob: Job? = null
        if (entry.terminalDeferred.isCompleted) {
            watcherJob = engineScope.launch(start = CoroutineStart.UNDISPATCHED) {
                val storedTerminal = entry.terminalDeferred.await() // returns immediately
                if (terminalFired.compareAndSet(false, true)) {
                    val terminalEnv: Map<String, Any?> = when {
                        storedTerminal === HUB_TERMINAL_OK ->
                            mapOf("ok" to true, "value" to null)
                        storedTerminal is HubTerminalError ->
                            mapOf(
                                "ok" to false,
                                "code" to "PROVIDER_ERROR",
                                "message" to storedTerminal.message,
                                "contractId" to contractId,
                                "member" to member,
                                "scope" to mapOf("kind" to scope.serialize()),
                            )
                        else -> mapOf("ok" to true, "value" to null)
                    }
                    onEnd(terminalEnv)
                    // Cancel the consumer coroutine so sharedFlow.collect stops.
                    consumerJob.cancel(CancellationException("h5-watcher"))
                }
            }
        }

        // Tie the hub refcount to the consumer job lifecycle. On completion or cancellation,
        // detach decrements the refcount and releases the upstream when the last consumer
        // leaves. Also cancels watcherJob to prevent leaks on consumer-cancel paths.
        consumerJob.invokeOnCompletion {
            watcherJob?.cancel()
            detach(contractId, member, scope, paramsHash)
        }

        // Return the consumer job so Router can track it in streamPumpJobs
        return consumerJob
    }

    /**
     * Detach a consumer. Decrements the refcount; cancels upstream + removes hub entry
     * when the last consumer leaves.
     */
    fun detach(contractId: String, member: String, scope: Scope, paramsHash: Long) {
        val key = HubKey(contractId, member, scope.serialize(), paramsHash)
        val entry = hubs[key] ?: return
        if (entry.refCount.decrementAndGet() <= 0) {
            entry.upstreamJob?.cancel()
            // Value-aware remove — only remove THIS entry so a live replacement
            // registered under the same key (new provider after cancel) is not evicted.
            hubs.remove(key, entry)
        }
    }

    /** Cancel all active hub entries (called on epoch swap). */
    fun cancelAll() {
        for (entry in hubs.values) {
            entry.upstreamJob?.cancel()
        }
        hubs.clear()
    }

    companion object {
        /** Sentinel object indicating normal stream completion from upstream. */
        internal val HUB_TERMINAL_OK: Any = object : Any() {}
    }

    internal data class HubTerminalError(val message: String)
}
